### PR TITLE
Generic pattern emission refactoring

### PR DESCRIPTION
**Pull request by slavapestov**
Concrete and generic metadata builders shared a base class, but ever since generic metadata patterns became true-const with a layout distinct from instantiated metadata, the generic case no longer uses most of the functionality of the base class.

This is a refactoring in preparation for giving concrete resilient classes a proper true-const pattern.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: lib/IRGen/GenMeta.cpp [hunk 1]
```diff
@@ -1532,27 +1532,204 @@ void IRGenModule::addFieldTypes(ArrayRef<CanType> fieldTypes) {
...
+
+  IGF.Builder.CreateLifetimeEnd(fields,
+                  IGF.IGM.getPointerSize() * storedProperties.size());
+}
+
+static void emitInitializeMetadata(IRGenFunction &IGF,
+                                   NominalTypeDecl *nominalDecl,
+                                   llvm::Value *metadata,
+                                   bool isVWTMutable,
+                                   MetadataDependencyCollector *collector) {
```
Comment (rjmccall) [inline lib/IRGen/GenMeta.cpp]: Should this be called `emitInitializeValueMetadata` to make it clear that it's not for classes?

**Review by rjmccall — APPROVED** [review]

File: lib/IRGen/GenMeta.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/IRGen/GenMeta.cpp]: Good idea

**Pull request merged**